### PR TITLE
Support building on Windows

### PR DESCRIPTION
**Pull request by kalrish**
This pull request includes the most basic and least invasive changes I could come up with to support building on Windows with MinGW. The build system is simplified along the way, with the Windows targets no longer being an extra, as they were before. A single source file is slightly modified.

Support for building on Windows is a necessary step for improving Tup's state on this platform. It also encourages contributions from people working on this system.

### CODE REVIEW

**Comment by gittup** [comment]
Thank you for working on this! I started to merge some of the commits, and cleaned up a few things so that I could still build correctly locally. In the process it's likely I broke the intended functionality of building locally on Windows (but unfortunately that didn't work for me with your branch, so I was unable to verify that I didn't further break things).

I had to update my test infrastructure to cross-compile using CONFIG_TARGET and CONFIG_CC/CC32 instead of the dual-build setup we had before, but now that it is setup it should be easier for me to verify further patches. I hope with the ones I've merged so far it will be easier for you to get the rest in a good state.

I'll try to comment on individual patches for reasons they weren't merged at the present time. Feel free to provide additional justification for them if they are needed.

File: src/dllinject/dllinject.c [hunk 1]
```diff
@@ -2187,4 +2197,4 @@ int tup_inject_dll(
 	}
 
 	return 0;
-}
+}
```
Comment (gittup) [inline src/dllinject/dllinject.c]: Can you please configure your editor to write out a newline at the end of the file? It breaks the diffs otherwise.

File: src/tup/Tupfile [hunk 2]
```diff
@@ -1,3 +1,10 @@
 include_rules
 : foreach *.c | ../luabuiltin/luabuiltin.h |> !cc |>
-: foreach *.c | ../luabuiltin/luabuiltin.h |> !mingwcc |>
+ifneq (@(TUP_PLATFORM),win32)
+: |> version=`git describe` ; echo "const char *tup_version(void) {return \"$version\";}" >version.c |> version.c
```
Comment (gittup) [inline src/tup/Tupfile]: I left only this rule in for now, even on Windows. I do not see the error you describe when I run it on my Windows machine, and we'll want to have an actual version number built-in. Does it help if the commands are moved into a shell script and then the rule is invoked as ': |> bash generate-version.bash > %o |> version.c' or something similar?

File: Tupfile [hunk 3]
```diff
@@ -21,10 +21,6 @@ else
 srcs += src/sqlite3/*.o
 endif
 ifneq ($(TARGET),win32)
-ifeq (@(TUP_SUDO_SUID),y)
```
Comment (gittup) [inline Tupfile]: There are still some systems that don't have namespacing support by default, so having suid support makes things easier I think. In any case, I don't think this has to do with Windows support, so it should probably be a separate pull request?

File: Tuprules.tup [hunk 4]
```diff
@@ -46,7 +46,7 @@ export PKG_CONFIG_PATH
 !ld = |> ^ LINK %o^ $(CC) %f -o %o $(LDFLAGS) $(LDFLAGS_%o) |>
 !ar = |> ^ AR %o^ ar crs %o %f |>
 !dot = |> ^ DOT %f^ dot -Efontname="Vernada, serif" -Nfontname="Vernada, serif" -Efontsize=10 -Nfontsize=10 -Tpng %f > %o |>
-!cp = |> ^ CP %f -> %o^ cp %f %o |>
+!cp = |> ^ CP %f -> %o^ cp -- %f %o |>
```
Comment (gittup) [inline Tuprules.tup]: What did you need this for? There still aren't any files with dashes, as far as I can see.

File: Tuprules.tup [hunk 5]
```diff
@@ -42,8 +42,8 @@ endif
 
 export PKG_CONFIG_PATH
 
-!cc = |> ^ CC %f^ $(CC) -c %f -o %o $(CFLAGS) $(CFLAGS_%f) |> %B.o
```
Comment (gittup) [inline Tuprules.tup]: Is this needed to support other compilers or something?

File: win32.tup [hunk 6]
```diff
@@ -43,6 +43,10 @@ CFLAGS += -D_UNICODE
 
 CFLAGS32 += $(CFLAGS)
 
+LDFLAGS += -static-libgcc
+
+LDFLAGS32 += $(LDFLAGS)
```
Comment (gittup) [inline win32.tup]: I had to change this to use $(LDFLAGS) in the rule rather than adding them to LDFLAGS32 here because it wasn't getting picked up in Tupfiles where we add to LDFLAGS after include_rules.

File: tup.config [hunk 7]
```diff
@@ -1,2 +1 @@
-CONFIG_CC32=gcc-x86_64-x86.bat
```
Comment (gittup) [inline tup.config]: It would be very helpful if you squashed commits like this with the ones that introduced the files in the first place. It will be easier for me to test and review.

File: Tuprules.tup [hunk 8]
```diff
@@ -12,38 +12,11 @@ else
...
-CFLAGS += -Os
-endif
+CFLAGS = @(CFLAGS)
 
-CFLAGS += -W
-CFLAGS += -Wall
-ifeq (@(TUP_WERROR),y)
-CFLAGS += -Werror
-endif
-CFLAGS += -Wbad-function-cast
```
Comment (gittup) [inline Tuprules.tup]: I somewhat like the idea of having different sensible configs in tree, but I'm not totally sold that a reasonable default doesn't belong in the Tupfiles. I think a reasonable default is the flags that I build with, since it is known to compile on a few platforms for at least some version of gcc.

**Comment by gittup** [comment]
That's all I remember for now - I should've taken better notes :/. Hopefully you are able to rebase and get it working. Thanks again for your help!

File: src/dllinject/dllinject.c (hunk 1 above)
Comment (kalrish) [inline src/dllinject/dllinject.c]: Done. Sorry.

File: src/tup/Tupfile (hunk 2 above)
Comment (kalrish) [inline src/tup/Tupfile]: Thankfully you have solved the issue with the bizarre output file with bdae36034b39e02f39931fe9f5ff2199cab4cb8c. It might have had something to do with the specific version of git I was using. Still, the rule uses syntax specific to Unix-style shells and doesn't work here (cmd.exe complains about invalid syntax), so an alternative for Windows seems necessary. I'll provide one soon.

**Comment by kalrish** [comment]
Thank you for having a look. With your latest commits, tup was just one patch away from building locally here on Windows.

This pull request included mixed changes because I forgot to commit what pertained to it to a separate branch, and instead just committed to master. Later, when I added more commits, they showed up here without me noticing. Sorry for the confusion.

I'll open a new, proper pull request shortly with the remaining changes related to building on Windows.

File: Tuprules.tup (hunk 8 above)
Comment (kalrish) [inline Tuprules.tup]: You are right. On the one hand, moving everything toolchain-specific to separate configs would allow to remove some ifdefs from Tupfiles. The sanctioned default could still be in tup.config so that this restructuration was transparent to users. But, on the other hand, having the default in tup.config would mean having it hard-coded, meaning that those cloning and building on Windows would get wrong flags by default. By having the default in Tupfiles, we are able to use conditional constructs to set sensible defaults depending on the target platform.

That said, @-variables like `CONFIG_TUP_WERROR` aren't needed. We don't really want to have a config variable for each compiler feature; it's just easier to let users specify those flags in their tup.config. `-Werror` and such should be moved to a template tup.config for developers (say, `dev.config`), while similar templates could be provided for release and LTO builds and whatnot.
